File: werkzeuge/freigabe.py
#!/usr/bin/env python3
"""
Was auf Firat wartet — Issues fuer Gate G2, PRs fuer eine Freigabe.

Am 11. und 12.09.2026 hat er rund dreissig Mal
"gh issue edit --add-label status:freigegeben" getippt. Das war die
tatsaechliche Reibung dieser Tage, mehr als jede Regel.

MIT WESSEN ZUGANG
-----------------
Ohne gesetztes GH_TOKEN, also mit Firats Anmeldung. Laeuft es ueber das
Bot-Token, steht in der Historie der Bot — und Gate G2 ist eine Attrappe.
Genau das ist am 2026-09-11 zweimal passiert (#10, #57). Deshalb gibt es
dafuer einen eigenen Test und nicht nur einen Kommentar.
"""

from __future__ import annotations

import json
import os
import re
import subprocess
from typing import Callable, Sequence

Aufruf = Callable[[Sequence[str]], str]

FERTIG_WENN = re.compile(r"##\s*Fertig,\s*wenn", re.IGNORECASE)


def _gh(befehl: Sequence[str]) -> str:
    """
    Ruft gh mit Firats Anmeldung.

    GH_TOKEN wird ausdruecklich aus der Umgebung genommen: Der Server laeuft
    unter Umstaenden neben einer Sitzung, die es fuer den Bot gesetzt hat.
    """
    umgebung = {k: v for k, v in os.environ.items() if k not in ("GH_TOKEN", "GITHUB_TOKEN")}
    ergebnis = subprocess.run(
        ["gh", *befehl], capture_output=True, text=True, env=umgebung, check=False
    )
    if ergebnis.returncode != 0:
        # Durchreichen, nicht verschlucken. Ein Fehler, den niemand sieht,
        # sieht aus wie Erfolg — das Muster dieses Projekts (#7, #9, #76).
        raise RuntimeError(ergebnis.stderr.strip() or "gh ist fehlgeschlagen")
    return ergebnis.stdout


def hat_abnahmekriterium(text: str) -> bool:
    """CLAUDE.md: Kein Issue ohne Abnahmekriterium. PR #68 ist daran gescheitert."""
    return bool(FERTIG_WENN.search(text or ""))


def offene_issues(aufruf: Aufruf = _gh) -> list[dict]:
    """Was auf Gate G2 wartet: vorgeschlagen, aber nicht freigegeben."""
    roh = json.loads(aufruf([
        "issue", "list", "--state", "open", "--limit", "100",
        "--json", "number,title,labels,body,milestone",
    ]) or "[]")

    offen = []
    for i in roh:
        namen = {l["name"] for l in i.get("labels", [])}
        if "status:freigegeben" in namen or "status:vorschlag" not in namen:
            continue
        offen.append({
            "nummer": i["number"],
            "titel": i["title"],
            "groesse": next((n for n in namen if n.startswith("groesse:")), None),
            "art": next((n for n in namen if n.startswith("art:")), None),
            "meilenstein": (i.get("milestone") or {}).get("title"),
            "kriterium": hat_abnahmekriterium(i.get("body", "")),
        })
    return offen


def offene_prs(aufruf: Aufruf = _gh) -> list[dict]:
    """Offene PRs mit dem Stand ihrer Checks."""
    roh = json.loads(aufruf([
        "pr", "list", "--state", "open", "--limit", "50",
        "--json", "number,title,body,reviews,statusCheckRollup,files",
    ]) or "[]")

    offen = []
    for p in roh:
        checks = p.get("statusCheckRollup") or []
        rot = sorted({c.get("name") for c in checks if c.get("conclusion") == "FAILURE"})
        freigegeben = any(r.get("state") == "APPROVED" for r in p.get("reviews", []))
        bezug = re.search(r"(?im)^\s*refs\s+#(\d+)", p.get("body") or "")
        offen.append({
            "nummer": p["number"],
            "titel": p["title"],
            "issue": int(bezug.group(1)) if bezug else None,
            "rote_checks": rot,
            "freigegeben": freigegeben,
            "dateien": [f["path"] for f in (p.get("files") or [])][:10],
        })
    return offen


def issue_freigeben(nummer: int, aufruf: Aufruf = _gh) -> None:
    """Gate G2. Nur ein Mensch darf das — deshalb ohne Bot-Token."""
    aufruf(["issue", "edit", str(nummer), "--add-label", "status:freigegeben"])


def pr_freigeben(nummer: int, aufruf: Aufruf = _gh) -> None:
    aufruf(["pr", "review", str(nummer), "--approve"])


def kommentieren(art: str, nummer: int, text: str, aufruf: Aufruf = _gh) -> None:
    if art not in ("issue", "pr"):
        raise ValueError(f"Weder Issue noch PR: {art!r}")
    aufruf([art, "comment", str(nummer), "--body", text])

File: werkzeuge/freigabe.js
// Darstellung der Freigabe-Konsole (#90).
//
// Zwei Listen, weil zwei verschiedene Entscheidungen dahinterstehen: Ein
// Issue freizugeben heisst "das soll gebaut werden". Einen PR freizugeben
// heisst "so, wie es gebaut wurde, ist es richtig".
//
// Die Knoepfe machen das Freigeben leichter — und damit auch das Freigeben
// ohne Hinsehen. Deshalb steht neben jedem Knopf das, was man wissen muss,
// um ihn nicht zu druecken: fehlendes Abnahmekriterium, rote Checks.

export function maskiert(text) {
  return String(text ?? "")
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

/** CLAUDE.md verlangt ein Abnahmekriterium. PR #68 ist daran gescheitert. */
function warnungOhneKriterium(issue) {
  return issue.kriterium
    ? ""
    : `<p class="warnung">Kein „Fertig, wenn …“ — nach CLAUDE.md unvollständig.` +
      ` <code>abnahme</code> wird jeden PR dazu blockieren.</p>`;
}

export function issueZuHtml(issue) {
  const marken = [issue.groesse, issue.art, issue.meilenstein]
    .filter(Boolean)
    .map((m) => `<span class="marke">${maskiert(m)}</span>`)
    .join("");
  return (
    `<article class="eintrag${issue.kriterium ? "" : " unvollstaendig"}" data-nummer="${issue.nummer}">` +
    `<p class="kopf"><span class="nummer">#${issue.nummer}</span> ${maskiert(issue.titel)}</p>` +
    `<p class="marken">${marken}</p>` +
    warnungOhneKriterium(issue) +
    `<div class="knoepfe">` +
    `<button type="button" data-tat="freigeben" data-art="issue" data-nummer="${issue.nummer}">Freigeben</button>` +
    `<button type="button" data-tat="kommentieren" data-art="issue" data-nummer="${issue.nummer}">Rückfrage</button>` +
    `</div></article>`
  );
}

/** Rote Checks beim Namen nennen. Ein Knopf ohne diese Angabe lädt zum Wegsehen ein (#5). */
function checkStand(pr) {
  if (pr.rote_checks?.length) {
    return `<p class="warnung">Rot: ${pr.rote_checks.map(maskiert).join(", ")}</p>`;
  }
  return `<p class="gut">Alle Checks grün</p>`;
}

export function prZuHtml(pr) {
  const bezug = pr.issue ? `<span class="marke">Refs #${pr.issue}</span>` :
    `<span class="marke fehlt">kein Issue verlinkt</span>`;
  const dateien = (pr.dateien ?? []).slice(0, 5).map(maskiert).join(", ");
  return (
    `<article class="eintrag${pr.rote_checks?.length ? " unvollstaendig" : ""}" data-nummer="${pr.nummer}">` +
    `<p class="kopf"><span class="nummer">#${pr.nummer}</span> ${maskiert(pr.titel)}</p>` +
    `<p class="marken">${bezug}${pr.freigegeben ? `<span class="marke">freigegeben</span>` : ""}</p>` +
    checkStand(pr) +
    (dateien ? `<p class="dateien">${dateien}</p>` : "") +
    `<div class="knoepfe">` +
    `<button type="button" data-tat="freigeben" data-art="pr" data-nummer="${pr.nummer}"${pr.freigegeben ? " disabled" : ""}>Freigeben</button>` +
    `<button type="button" data-tat="kommentieren" data-art="pr" data-nummer="${pr.nummer}">Rückfrage</button>` +
    `</div></article>`
  );
}

export function konsoleZuHtml(daten) {
  const issues = daten.issues ?? [];
  const prs = daten.prs ?? [];
  return (
    `<section><h2>Warten auf Freigabe (${issues.length})</h2>` +
    (issues.length ? issues.map(issueZuHtml).join("") : `<p class="leer">Nichts offen.</p>`) +
    `</section>` +
    `<section><h2>Pull Requests (${prs.length})</h2>` +
    (prs.length ? prs.map(prZuHtml).join("") : `<p class="leer">Nichts offen.</p>`) +
    `</section>`
  );
}
